### PR TITLE
[clang]  Reland Add tanf16 builtin and support for tan constrained intrinsic

### LINKED ISSUE #91421: [HLSL] Add experimental_constrained_tan intrinsic
We want to do the same thing that cos\sin do in the below examples:

```cpp
  case Builtin::BI__builtin_cos:
  case Builtin::BI__builtin_cosf:
  case Builtin::BI__builtin_cosf16:
  case Builtin::BI__builtin_cosl:
  case Builtin::BI__builtin_cosf128:
    return RValue::get(emitUnaryMaybeConstrainedFPBuiltin(*this, E,
                                 Intrinsic::cos,
                                 Intrinsic::experimental_constrained_cos));
  
  case Builtin::BI__builtin_sin:
  case Builtin::BI__builtin_sinf:
  case Builtin::BI__builtin_sinf16:
  case Builtin::BI__builtin_sinl:
  case Builtin::BI__builtin_sinf128:
    return RValue::get(emitUnaryMaybeConstrainedFPBuiltin(*this, E,
                                 Intrinsic::sin,
                                 Intrinsic::experimental_constrained_sin));

```

_Originally posted by @farzonl in https://github.com/llvm/llvm-project/issues/90276#issuecomment-2083588945_

### PR DESCRIPTION
**Pull request by farzonl**
Relanding this PR now that https://github.com/llvm/llvm-project/pull/90503 has merged. with `FTAN` landing in [TargetLoweringBase.cpp:L1021](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/TargetLoweringBase.cpp#L1020C23-L1021C63 ) There is now a llvm tan intrinsic 32\64\128 Expand case for all  llvm backends. 

In LLVM, the `llvm.experimental.constrained.cos` and `llvm.experimental.constrained.sin` intrinsics are used for performing cosine and sine calculations with additional constraints on floating-point operations.  This behavior is expected for all floating-point math intrinsics. This change adds these constraints for the `tan` intrinsic.

-  `Builtins.td` - replace TanF128 with F16F128MathTemplate
-  `CGBuiltin.cpp` - map existing tan builtins to  `tan` and `constrained_tan` intrinsic
-   `ConstrainedOps.def` map tan and constrained_tan  to an ISDOpcode.

resolves  #91421

### CODE REVIEW

**Comment by farzonl** [comment]
@ilovepi  I minimised the failing issue from: https://fuchsia.googlesource.com/fuchsia/+/refs/heads/main/src/graphics/lib/compute/spinel/ext/transform_stack/transform_stack.c
to:
```cpp
#include <math.h>
#include <stdio.h>

#define SPN_TRANSFORM_STACK_TAN(x_)  tanf(x_)

float
spinel_transform_stack_push_skew_x(float theta)
{
  float const tan_theta = SPN_TRANSFORM_STACK_TAN(theta);
  return tan_theta;
}

int main() {
    printf("%f\n", spinel_transform_stack_push_skew_x(1.05f));
    return 0;
}
```
run like so:
```bash
farzon@devbox: projects/llvm-project$ <build_dir>/bin/clang  <test_path>/test.c -lm -o  <test_out_path>/test.out
farzon@devbox: projects/llvm-project$ ./<test_out_path>/test.out 
1.743315
```

The intrinsic should be able to be lowered now.

**Comment by bob80905** [comment]
Should a test be added when the return type is a different float type than the first arg type?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -2923,6 +2923,18 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       SetSqrtFPAccuracy(Call);
       return RValue::get(Call);
     }
+
+    case Builtin::BItan:
```
Comment (bob80905) [inline clang/lib/CodeGen/CGBuiltin.cpp:2927]: What is the difference between `BItanf` and `BI_builtin_tanf` ?

**Comment by ilovepi** [comment]
> @ilovepi I minimised the failing issue from: https://fuchsia.googlesource.com/fuchsia/+/refs/heads/main/src/graphics/lib/compute/spinel/ext/transform_stack/transform_stack.c to:
> 
> ```c++
> #include <math.h>
> #include <stdio.h>
> 
> #define SPN_TRANSFORM_STACK_TAN(x_)  tanf(x_)
> 
> float
> spinel_transform_stack_push_skew_x(float theta)
> {
>   float const tan_theta = SPN_TRANSFORM_STACK_TAN(theta);
>   return tan_theta;
> }
> 
> int main() {
>     printf("%f\n", spinel_transform_stack_push_skew_x(1.05f));
>     return 0;
> }
> ```
> 
> run like so:
> 
> ```shell
> farzon@devbox: projects/llvm-project$ <build_dir>/bin/clang  <test_path>/test.c -lm -o  <test_out_path>/test.out
> farzon@devbox: projects/llvm-project$ ./<test_out_path>/test.out 
> 1.743315
> ```
> 
> The intrinsic should be able to be lowered now.

Thanks. This seems fine to me. I'll defer to the other reviewers on everything else, but I'm satisfied with the new test. so LGTM from my perspective.

**Comment by farzonl** [comment]
> Should a test be added when the return type is a different float type than the first arg type?

So  if our function that calls the builtin has a different return type than the input that we pass in to the builtin then that will be handled by other casting code that has its own tests so no.

Second way this could happen if tan had different return type than from input type but in this case input dictates return type for tan.  So the answer again would be no.

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 1 above)
Comment (farzonl) [reply clang/lib/CodeGen/CGBuiltin.cpp:2927]: def TanF16F128 : Builtin, F16F128MathTemplate = BI_builtin_tanf
vs 
def Tan : FPMathTemplate, LibBuiltin<"math.h">  = BItanf

File: llvm/test/CodeGen/X86/vector-constrained-fp-intrinsics.ll [hunk 2]
```diff
@@ -7881,7 +8111,7 @@ declare <3 x double> @llvm.experimental.constrained.powi.v3f64(<3 x double>, i32
 declare <3 x float> @llvm.experimental.constrained.sin.v3f32(<3 x float>, metadata, metadata)
 declare <3 x double> @llvm.experimental.constrained.sin.v3f64(<3 x double>, metadata, metadata)
 declare <3 x float> @llvm.experimental.constrained.cos.v3f32(<3 x float>, metadata, metadata)
-declare <3 x double> @llvm.experimental.constrained.cos.v3f64(<3 x double>, metadata, metadata)
```
Comment (farzonl) [inline llvm/test/CodeGen/X86/vector-constrained-fp-intrinsics.ll:7884]: typo:
declare <3 x float> @llvm.experimental.constrained.cos.v3f32(<3 x float>, metadata, metadata)
declare <3 x double> @llvm.experimental.constrained.cos.v3f64(<3 x double>, metadata, metadata)

AND

declare <3 x float> @llvm.experimental.constrained.tan.v3f32(<3 x float>, metadata, metadata)
declare <3 x double> @llvm.experimental.constrained.tan.v3f64(<3 x double>, metadata, metadata)

File: llvm/test/Feature/fp-intrinsics.ll [hunk 3]
```diff
@@ -151,6 +151,17 @@ entry:
...
+; CHECK-LABEL: ftan
+; CHECK: call double @llvm.experimental.constrained.tan
+define double @ftan() #0 {
+entry:
+  %result = call double @llvm.experimental.constrained.tan.f64(double 42.0,
+                                               metadata !"round.dynamic",
+                                               metadata !"fpexcept.strict") #0
+  ret double %result
+}
+
```
Comment (farzonl) [inline llvm/test/Feature/fp-intrinsics.ll:164]: add the @llvm.experimental.constrained.tan.f64 declare

**Comment by farzonl** [comment]
@efriedma-quic If you have time could you take a look at this pr. It  is the same as the one you review just with more tests across more targets.  Problem from the first merge was because the  target base changes were tied up in the x86 backend change.  Since those have merged this pr should work just  fine now.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by glandium** [comment]
This broke building wasi-libc, with a crash:
```
Stack dump:
0.	Program arguments: /builds/worker/fetches/clang/bin/clang-19 -cc1 -triple wasm32-unknown-wasi -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name ctanh.c -mrelocation-model static -mthread-model single -mframe-pointer=none -ffp-contract=on -fno-rounding-math -ffp-exception-behavior=ignore -mconstructor-aliases -target-cpu generic -fvisibility=hidden -debugger-tuning=gdb -fdebug-compilation-dir=/builds/worker/fetches/wasi-sdk/src/wasi-libc -fcoverage-compilation-dir=/builds/worker/fetches/wasi-sdk/src/wasi-libc -sys-header-deps -D NDEBUG -O2 -Wall -Wextra -Wno-null-pointer-arithmetic -Wno-unused-parameter -Wno-sign-compare -Wno-unused-variable -Wno-unused-function -Wno-ignored-attributes -Wno-missing-braces -Wno-ignored-pragmas -Wno-unused-but-set-variable -Wno-unknown-warning-option -Wno-parentheses -Wno-shift-op-parentheses -Wno-bitwise-op-parentheses -Wno-logical-op-parentheses -Wno-string-plus-int -Wno-dangling-else -Wno-unknown-pragmas -ferror-limit 19 -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -vectorize-loops -vectorize-slp -x c ctanh-46a6c8.c
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'ctanh-46a6c8.c'.
4.	Running pass 'WebAssembly Assembly Printer' on function '@ctanh'
 #0 0x00007f2205ba35cd llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /tmp/llvm/llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x00007f2205ba3abb PrintStackTraceSignalHandler(void*) /tmp/llvm/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x00007f2205ba1b26 llvm::sys::RunSignalHandlers() /tmp/llvm/llvm/lib/Support/Signals.cpp:105:5
 #3 0x00007f2205ba4275 SignalHandler(int) /tmp/llvm/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007f2203ff4f90 (/lib/x86_64-linux-gnu/libc.so.6+0x3bf90)
 #5 0x00007f2204043ccc (/lib/x86_64-linux-gnu/libc.so.6+0x8accc)
 #6 0x00007f2203ff4ef2 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3bef2)
 #7 0x00007f2203fdf472 abort (/lib/x86_64-linux-gnu/libc.so.6+0x26472)
 #8 0x00007f2205a7f300 llvm::install_out_of_memory_new_handler() /tmp/llvm/llvm/lib/Support/ErrorHandling.cpp:194:0
 #9 0x00007f220ba5052d llvm::WebAssembly::getLibcallSignature(llvm::WebAssemblySubtarget const&, llvm::StringRef, llvm::SmallVectorImpl<llvm::wasm::ValType>&, llvm::SmallVectorImpl<llvm::wasm::ValType>&) /tmp/llvm/llvm/lib/Target/WebAssembly/WebAssemblyRuntimeLibcallSignatures.cpp:907:30
#10 0x00007f220b9c4e0f llvm::WebAssemblyAsmPrinter::getOrCreateWasmSymbol(llvm::StringRef) /tmp/llvm/llvm/lib/Target/WebAssembly/WebAssemblyAsmPrinter.cpp:0:5
#11 0x00007f220b9c4f46 llvm::WebAssemblyAsmPrinter::emitDecls(llvm::Module const&) /tmp/llvm/llvm/lib/Target/WebAssembly/WebAssemblyAsmPrinter.cpp:310:40
#12 0x00007f220b9c7179 llvm::WebAssemblyAsmPrinter::emitConstantPool() /tmp/llvm/llvm/lib/Target/WebAssembly/WebAssemblyAsmPrinter.cpp:604:3
#13 0x00007f2206ee9526 llvm::AsmPrinter::emitFunctionHeader() /tmp/llvm/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:966:3
```
This file contains the script and preprocessed source that reproduce the problem: [ctanh.zip](https://github.com/user-attachments/files/15783347/ctanh.zip)

**Comment by aeubanks** [comment]
I believe this has exposed a preexisting issue in isel with tan:

```
$ cat /tmp/b.ll
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7-unknown-linux-android26"

define <2 x float> @g() {
entry:
  %0 = call <2 x float> @llvm.tan.v2f32(<2 x float> zeroinitializer)
  ret <2 x float> %0
}
$ llc /tmp/b.ll -o /dev/null
LLVM ERROR: Cannot select: t3: v2f32 = ftan t15                                                                                                                                                                                                                                                                    
  t15: v2f32 = bitcast t14                                                                                                                                                                                                                                                                                         
    t14: v2i32 = ARMISD::VMOVIMM TargetConstant:i32<0>                                                                                                   
      t13: i32 = TargetConstant<0>                                                                                                                                                                                                                                                                                 
```

**Comment by farzonl** [comment]
@aeubanks The issue you are seeing here is because only aarch64 and x86 backends have tan intrinsic lowering support.  This shouldn't be a regression because there never was a llvm.tan.v2f32 for ARMv7.  That still needs to be add. Can you show me how you could have  generated `llvm.tan.v2f32` for armv7 from clang?

By default [TargetLoweringBase.cpp:1016-1022](https://github.com/llvm/llvm-project/blob/8f6acd973a38da6dce45faa676cbb51da37f72e5/llvm/lib/CodeGen/TargetLoweringBase.cpp#L1016C3-L1022C1) only supports lowering form f32, f64, and f128 scalars across all backends.

The builtins that were added (`Builtin::BItanf`, `Builtin::BItanl`, `Builtin::BI__builtin_tan`, `Builtin::BI__builtin_tanf`, `Builtin::BI__builtin_tanf16`, `Builtin::BI__builtin_tanl`, and `Builtin::BI__builtin_tanf128`) should only cover scalar cases.

So i'm not sure how this pr could have generate the intrinsic you have reported.

**Comment by aeubanks** [comment]
SLPVectorizer can introduce `llvm.tan.v2f32`. For example, running `opt -O3` on the following introduces `llvm.tan.v2f32`
```
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7-unknown-linux-android26"

; Function Attrs: null_pointer_is_valid
define void @foo() #0 {
bb:
  %call = call float @pluto(float 1.000000e+00)
  %call1 = call i1 @ham(float %call)
  %call2 = call float @pluto(float 0.000000e+00)
  %call3 = call i1 @ham(float %call2)
  %select = select i1 %call3, float 0.000000e+00, float 1.000000e+00
  %select4 = select i1 %call1, float 0.000000e+00, float 1.000000e+00
  store float %select, ptr null, align 4
  %call5 = call ptr null(ptr null, float %select4, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00)
  ret void
}

define float @pluto(float %arg) {
bb:
  %call = call float @llvm.tan.f32(float %arg)
  ret float %call
}

define i1 @ham(float %arg) {
bb:
  %fcmp = fcmp ole float %arg, 0.000000e+00
  ret i1 %fcmp
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare float @llvm.tan.f32(float) #1

attributes #0 = { null_pointer_is_valid }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

**Comment by farzonl** [comment]
It seems like we have four options here. We can drop  the `def Tan : FPMathTemplate, LibBuiltin<"math.h">` builtins so no  `Builtin::BItanf` or `Builtin::BItanl` in cgbuitlin switch case. That wouldn't solve the SLPVectorizer  case  but doesn't expose it either unless you use a clang_builtin function which should limit the impact. The tan library functions would behave as before.

Option 2 we land an arm7 backend for Tan.  That could work for you, but the issue you raised with the `SLPVectorizer` case likely means this same issue exists for powerPC, RISCV, and potentially other backends.

Option 3.  we guard the  emitter with a target check for wasm (maybe), x86, and aarch64
```cpp
    case Builtin::BItanf:
    case Builtin::BItanl:
    case Builtin::BI__builtin_tan:
    case Builtin::BI__builtin_tanf:
    case Builtin::BI__builtin_tanf16:
    case Builtin::BI__builtin_tanl:
    case Builtin::BI__builtin_tanf128: {
      switch(CGF.getTarget().getTriple().getArch()) {
           case llvm::Triple::aarch64:
           case llvm::Triple::x86:
           case llvm::Triple::x86_64:
           case llvm::Triple::wasm32:
           case llvm::Triple::wasm64:
               return RValue::get(emitUnaryMaybeConstrainedFPBuiltin(
                      *this, E, Intrinsic::tan, Intrinsic::experimental_constrained_tan));
      }
}
```

Option 4 we revert this change, The test cases would go stale and it also means our plans for landing constraint intrinsics needs to go in the backlog, because it means constraint intrinsics don't make sense until tan has full support across all backends.

**Comment by efriedma-quic** [comment]
Can we change the target-independent bits of the tan() implementation in the backend so it doesn't require each target to explicitly request that tan() needs to be expanded?  It should be possible to adjust the code in TargetLoweringBase.cpp a bit so FTAN defaults to being expanded for all types.

**Comment by davemgreen** [comment]
If you remove tan from isTriviallyVectorizable it should prevent vectorization in the short term.

It might be better to default FTAN to expand in https://github.com/llvm/llvm-project/blob/64c9a1e1266ec7bc4c4896b2df116fa12dbacf15/llvm/lib/CodeGen/TargetLoweringBase.cpp#L960, which seems to only be done for f32/f64/f128 at the moment.

**Comment by farzonl** [comment]
> Can we change the target-independent bits of the tan() implementation in the backend so it doesn't require each target to explicitly request that tan() needs to be expanded? It should be possible to adjust the code in TargetLoweringBase.cpp a bit so FTAN defaults to being expanded for all types.

> It might be better to default FTAN to expand in
> 
> https://github.com/llvm/llvm-project/blob/64c9a1e1266ec7bc4c4896b2df116fa12dbacf15/llvm/lib/CodeGen/TargetLoweringBase.cpp#L960
> 
> , which seems to only be done for f32/f64/f128 at the moment.

I don't think we can default on vector types to Expand as it would expand on vector that are not supported for tan. For example VT in this case is `MVT::all_valuetypes()` That includes integer vectors.
 https://github.com/llvm/llvm-project/blob/64c9a1e1266ec7bc4c4896b2df116fa12dbacf15/llvm/lib/CodeGen/TargetLoweringBase.cpp#L894

Even if it only included floats it would not necessarily be correct For example in ARM
https://github.com/llvm/llvm-project/blob/64c9a1e1266ec7bc4c4896b2df116fa12dbacf15/llvm/lib/Target/ARM/ARMISelLowering.cpp#L324-L325
some cases like MVE  it is MVT::v8f16, MVT::v4f32
but for NEON MVT::v2f64, MVT::v4f32, and MVT::v2f32 are supported. Doing it per target seems intential so as to have more precision so only intrinsics that are valid to lower get expanded.

**Comment by davemgreen** [comment]
Usually when new ISD nodes are added they are expanded for all types, so that every backend will get at least working code even if it is not optimal. The targets can then come along and override the defaults for the types they are interested in, to get better results.

For tan I would expect most vector types would want to scalarize, so marking them as expand would make sense. If more types than are necessary get marked as Expand that shouldn't be an issue, it looks like we already do that for a number of other nodes.

**Comment by aeubanks** [comment]
@alexey-bataev

**Comment by farzonl** [comment]
> Usually when new ISD nodes are added they are expanded for all types, so that every backend will get at least working code even if it is not optimal. The targets can then come along and override the defaults for the types they are interested in, to get better results.
> 
> For tan I would expect most vector types would want to scalarize, so marking them as expand would make sense. If more types than are necessary get marked as Expand that shouldn't be an issue, it looks like we already do that for a number of other nodes.

 Should this be considered a termporary change until their is parity across alll backends? I'm trying to understand why no on else has done this for sin\cos\log\exp\etc.

**Comment by davemgreen** [comment]
I believe they were added so long ago that the default Expanding wasn't done at the time. @efriedma-quic do you have more of an idea than that?

**Comment by efriedma-quic** [comment]
Yes, I think it's just a historical mistake; sin/cos/log/exp were added a very long time ago, and we weren't as careful about that sort of thing.  And nobody has taken the time to try to cleanup the current defaults.

**Comment by farzonl** [comment]
PR https://github.com/llvm/llvm-project/pull/95518/files should address all backends.

Also SLPVectorizer was disabled for tan so I put out PR https://github.com/llvm/llvm-project/pull/95517

**Pull request merged**